src/cli.rs

```diff
@@ -1,22 +1,120 @@
 //use web3::{Web3, transports};
+use clap::{App, Arg};
 use reqwest::blocking as reqwest;
 use serde_json;
+use std::env;
+use std::ffi::OsString;
 use std::fs;
 use std::io::{self, Read, Write};
+use std::net::TcpListener;
+use std::net::TcpStream;
 use std::os::unix;
 use threadpool::ThreadPool;
 
-// TODO: things to configure:
-//  - infura project id (not just env var?)
-//  - rpc endpoint port
-//  - rpc endpoint type (tcp, ws, ipc)
-//  - max concurrent requests (ie~ threadpool size)
+#[derive(Debug, PartialEq)]
+pub struct TrinConfig {
+    pub protocol: String,
+    pub infura_project_id: String,
+    pub endpoint: u32,
+    pub pool_size: u32,
+}
+
+impl TrinConfig {
+    pub fn new() -> Self {
+        Self::new_from(std::env::args_os().into_iter()).unwrap_or_else(|e| e.exit())
+    }
+
+    fn new_from<I, T>(args: I) -> Result<Self, clap::Error>
+    where
+        I: Iterator<Item = T>,
+        T: Into<OsString> + Clone,
+    {
+        let matches = App::new("trin")
+            .version("0.0.1")
+            .author("carver")
+            .about("super lightweight eth portal")
+            .arg(
+                Arg::with_name("protocol")
+                    .short("p")
+                    .long("protocol")
+                    .help("select transport protocol")
+                    .takes_value(true)
+                    .default_value("http"),
+            )
+            .arg(
+                Arg::with_name("endpoint")
+                    .short("e")
+                    .long("endpoint")
+                    .help("http port")
+                    .takes_value(true)
+                    .default_value("7878"),
+            )
+            .arg(
+                Arg::with_name("pool_size")
+                    .short("s")
+                    .long("pool-size")
+                    .help("max size of threadpool")
+                    .takes_value(true)
+                    .default_value("2"),
+            )
+            .get_matches_from_safe(args)
+            .unwrap_or_else(|e| panic!("Unable to parse args: {}", e));
+
+        println!("Launching Trin...");
+        let protocol = matches.value_of("protocol").unwrap();
+        let endpoint = matches.value_of("endpoint").unwrap();
+        let endpoint = match endpoint.parse::<u32>() {
```

Comment:
If I understand correctly, then in the case that ipc protocol is selected - we don't care to check a `http_port` argument might also be incorrectly provided? It makes sense - after all the name `http_port` is pretty indicative that it's incompatible with ipc protocol.

Reply:
I found a clap macro (`value_t`) that simplifies the code parsing/casting the cli args. How strongly do you feel about this? I see the benefit of a constant.. But, I've also tried a couple variants and haven't found a reasonably simple solution (but, my rust type-fu is weak so who knows what obvious solutions i'm missing). 

- since `TrinConfig` has defined `http_port: u32`, we eventually have to cast a `const http_port: String = "8545"` into an int when we use the default constant (in both the http & ipc cases)
- the `default_arg()` param for the clap parser only accepts strings, so we have to do some casting if `const http_port: u32 = 8545`
- A performance benefit is somewhat irrelevant if a user selects `ipc` and a custom `http_port` since we error out. In the case that the user just selects `ipc`, we will still have to cast a `const http_port: String = "8545"` into a string to feed into `TrinConfig`

I'm happy to keep tinkering with this, just reporting my initial findings